Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch

```diff
@@ -0,0 +1,76 @@
+From 288b2d33b5f0d069d52f08415a7389823f3c088e Mon Sep 17 00:00:00 2001
+From: miclebrick <miclebrick@outlook.com>
+Date: Thu, 23 Aug 2018 11:45:32 -0400
+Subject: [PATCH] Cached IBlockData to CraftBlockData conversion
+
+
+diff --git a/src/main/java/net/minecraft/server/BlockData.java b/src/main/java/net/minecraft/server/BlockData.java
+index 4dd6c3276..3449be6fe 100644
+--- a/src/main/java/net/minecraft/server/BlockData.java
++++ b/src/main/java/net/minecraft/server/BlockData.java
+@@ -1,6 +1,7 @@
+ package net.minecraft.server;
+ 
+ import com.google.common.collect.ImmutableMap;
++import org.bukkit.craftbukkit.block.data.CraftBlockData;
+ 
+ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements IBlockData {
+ 
+@@ -11,4 +12,14 @@ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements I
+     public Block getBlock() {
+         return (Block) this.e_;
+     }
++
++    // Paper start - impl cached craft block data, lazy load to fix issue with loading at the wrong time
++    private CraftBlockData cachedCraftBlockData;
++
++    @Override
++    public CraftBlockData getCachedCraftBlockData() {
++        if(cachedCraftBlockData == null) cachedCraftBlockData = java.util.Objects.requireNonNull(CraftBlockData.createData(this));
++        return cachedCraftBlockData;
++    }
++    // Paper end
+ }
+diff --git a/src/main/java/net/minecraft/server/IBlockData.java b/src/main/java/net/minecraft/server/IBlockData.java
+index b57f6efb3..df2db8721 100644
+--- a/src/main/java/net/minecraft/server/IBlockData.java
++++ b/src/main/java/net/minecraft/server/IBlockData.java
+@@ -28,6 +28,8 @@ public interface IBlockData extends IBlockDataHolder<IBlockData> {
+ 
+     Block getBlock();
+ 
++    org.bukkit.craftbukkit.block.data.CraftBlockData getCachedCraftBlockData(); // Paper - property for cached API BlockData to clone from, should only be used by CraftBlockData#fromData
++
+     default Material getMaterial() {
+         return this.getBlock().n(this);
+     }
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+index 472e89c43..705f8d08e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+@@ -454,6 +454,7 @@ public class CraftBlockData implements BlockData {
+         register(net.minecraft.server.BlockWitherSkull.class, org.bukkit.craftbukkit.block.impl.CraftWitherSkull.class);
+         register(net.minecraft.server.BlockWitherSkullWall.class, org.bukkit.craftbukkit.block.impl.CraftWitherSkullWall.class);
+         register(net.minecraft.server.BlockWoodButton.class, org.bukkit.craftbukkit.block.impl.CraftWoodButton.class);
++        Block.REGISTRY_ID.iterator().forEachRemaining(IBlockData::getCachedCraftBlockData); // Paper - initialize cached data for all IBlockData instances after registration
```

Comment:
Is this even necessary? Couldn't you just let it be lazily initialized the first time it is needed?

Reply:
That'd be slower when plugins are calling it during the run of the server, no reason *not* to

Reply:
why even do this? Initialize it in the constructor of BlockData like you did the previous pass.

Reply:
see discord for reasoning that broke it